### DISCUSSION TITLE
Why are lazy-loaded client component code sent to the browser, even though the component isn't used?

### DISCUSSION THREAD

**Discussion by gustavotoyotarh** [vercel/next.js#70512] [Help]
### Summary

I have a list of lazy components on my app, like this:

```
import dynamic from "next/dynamic";

const MyComponent1 = dynamic(() => import("./my-component-1"));
const MyComponent2 = dynamic(() => import("./my-component-2"));
const MyComponent3 = dynamic(() => import("./my-component-3"));

export const components: Record<string, any> = {
  MyComponent1,
  MyComponent2,
  MyComponent3
};
```

When I use one of the components on this list, I see that all the other components' client code are sent to the browser as well:

```
import { components } from "./components";

export default async function Home() {
  const Component = components["MyComponent1"];

  return (
    <div>
      <Component />
    </div>
  );
}
```

![Screenshot from 2024-09-26 10-21-28](https://github.com/user-attachments/assets/482c312a-db0b-45b1-8af6-00af0573b92e)

This is in production.

What is happening here?
What can I do to not send unused lazy components' code to the browser, but still allow them to be dynamically loaded?

**Comment by gustavotoyotarh** [comment]
Seems to be related to https://github.com/vercel/next.js/issues/49742

**Reply by gustavotoyotarh** [reply]
Also https://github.com/vercel/next.js/issues/50285

**Comment by icyJoseph** [comment]
What next.js version are you using? I had seen this issue be reported before, so I setup a small reproduction repo, and I can't get it to happen. Got a link to a repository I could see?

**Reply by gustavotoyotarh** [reply]
I'm using NextJS 14.2.13

Reproduction repo: https://github.com/gustavotoyotarh/unused-lazy-components-still-load-nextjs

**Reply by gustavotoyotarh** [reply]
When running a production build you can see that the unused lazy client components are sent to the browser:
![Screenshot from 2024-09-26 11-29-22](https://github.com/user-attachments/assets/39140eb2-f827-4133-8f02-6e08b4be8ea4)

**Reply by icyJoseph** [reply]
Ah.. I forgot to add `use client` to one of them...

**Reply by gustavotoyotarh** [reply]
Are you aware of any solution to this problem? The thing is that in my project all of these components are required to be in the source code, just different users will see a different set of these components, and I don't want to send all client components to all users.

**Reply by icyJoseph** [reply]
I have seen this before, but IIRC, this had something to do with the loader being eagerly executed:

- https://github.com/vercel/next.js/blob/3ed9f4b3f4d1f8b431ff04d4e6a45a949680a31f/packages/next/src/shared/lib/dynamic.tsx#L125C1-L129C1

```jsx
  const loader = () =>
    loaderFn != null
      ? loaderFn().then(convertModule)
      : Promise.resolve(convertModule(() => null))

```

I am going of memory here, but the loaderFn, is what you pass to the `dynamic` function, and as part of its function body, `dynamic`, eagerly evaluates the loaderFn.

Trying to find the discussion, but this was a couple of years ago..

**Reply by gustavotoyotarh** [reply]
The same thing happens when using React.lazy, though.

**Reply by gustavotoyotarh** [reply]
We are losing 24 points in the PageSpeed score of our website due to this problem, it has become quite a burden.

**Reply by icyJoseph** [reply]
> The same thing happens when using React.lazy, though.

Probably the same mechanism internally.

**Reply by birnayt** [reply]
@icyJoseph did you find the discussion about that topic?
